### PR TITLE
Update publication repository to retrieve random publications

### PR DESCRIPTION
**Pull request by MathieuxC**
This PR slightly changes the way PublicationRepository retrieve the publications from firestore. Instead of loading the 20 oldest one, it now loaded 20 on a fully random way. There was no great way of doing this as firestore doesn't have a built method to get some documents randomly so, after exploring many possibilities, I decided to make it charge all publlications and take some randomly.

### CODE REVIEW

**Review by AKallala — CHANGES_REQUESTED** [review]
I think that if the loadRandomPublications retrieves all the publications from the database without a limit, it could eventually exhaust the app's cache storage (since Firebase Firestore caches all accessed documents locally) at some point when the number of publications becomes very large

**Comment by MathieuxC** [comment]
> I think that if the loadRandomPublications retrieves all the publications from the database without a limit, it could eventually exhaust the app's cache storage (since Firebase Firestore caches all accessed documents locally) at some point when the number of publications becomes very large

Unfortunately firebase doesn't come with a way of retrieving random publications. To solve this issue I thought of several possibility :

1. Add a random field to the firestore document. Then when trying to get publications we would have to sort the database field, generate a new random number, and take the 20 upcomming publications. This caused a problem of backward compatibility as all the publications we have now would have to be deleted as they wouldn't work.
2. In the same idea but without that last problem, I thought of sorting the database by creation time, generating a number between the current time and the time of the first publication. This doesn't work as it would cause obvious issue if there is a big gab between some publication (e.g. we spend 1 month without posting anything).
3. Still in the same idea, but sorting the database by id. As they are generated pseudo-randomly this could be a possibility but isn't optimal as the chance of getting a publication wouldn't be equivalent for all (for example the "last" publications would be retrieved more that the "first" ones, and if some publication get a realy "low" id they would never be seen). Also all of those possiblity have a common downside that the chunks of publications are always the same.

Even if I am aware that this implementation wouldn't scale to a large amount of publications, I don't see any workaround to avoid that problem (except maybe changing of database but this won't happen now). As the time needed for the request is largely upperbounded by the travel time for the current size of our publications collection and as it shouldn't have any downside in the scope if a swent project I thought this was an acceptable solution. If you disagree or have any idea of a better way of doing thing I am of course open to suggestions.

Also as I mentioned in the meeting this way we could potentialy implement a suggestion algorithm on the application side, as it would not be possible on the server side unless we created our own servers.

File: app/src/main/java/com/github/se/eduverse/repository/PublicationRepository.kt
Comment (mehdixboutaleb) [inline app/src/main/java/com/github/se/eduverse/repository/PublicationRepository.kt:1]: LGTM, the fact of not using timestamp will make it random solving the issue

File: app/src/test/java/com/github/se/eduverse/repository/PublicationRepositoryTest.kt
Comment (mehdixboutaleb) [inline app/src/test/java/com/github/se/eduverse/repository/PublicationRepositoryTest.kt:1]: LGTM, good adaptation of the test to make the new code pass.

**Review by mehdixboutaleb — APPROVED** [review]
LGTM, I saw the discussion earlier with Ahmed, you have tried different implementation but there is something that doesn't pass for each of the solution you proposed, this implementation for now is enough. We will try to change this on the future.

**Comment by AKallala** [comment]
I think that the issue you're describing in 1. :
> 1. This caused a problem of backward compatibility as all the publications we have now would have to be deleted as they wouldn't work.

is not a significant problem for us. The current publications in the database are test data generated by the team and not actual user-generated content, so discarding them wouldn't be a big deal. Additionally, we don't necessarily need to delete the old publications; they can simply be updated using a firebase.batch.update() operation since the current dataset is small.
Regarding the random publications display, I believe the first solution (adding a randomNbr field to publications) is the best option. However, it has a potential drawback: when loadMorePublications() is called in the ViewModel, the order of publications will completely change. This means that if a user scrolls up, they won't see the previous publications, which might feel inconsistent.
We need to decide whether to:
1. Provide a linear, consistent publications feed where scrolling back shows the same content.
2. Use a dynamic, randomized feed where scrolling back up may not return to previously viewed publications. 

I think that the latter is simpler, more practical and aligns well with the « content discovery hub" concept of the TikTok-like screen. If users want to revisit a publication, they can still add it to their favorites or save it. However, if we choose this approach, we should find a way to inform users about this behavior to set the right expectations.

**Review by AKallala — APPROVED** [review]
lgtm

**Pull request merged**